### PR TITLE
MAINT: refactor PyArrayMultiIterObject constructors

### PR DESCRIPTION
**Pull request by mattip**
Creates a single private implementation of the `PyArrayMultiIterObject`
constructor, and calls it from the three existing public constructors.

Rebased #7441 off master and made a new PR since I could not push the rebase to @jaimefrio 's repo. Of course we only need one of these, @jaimefrio if you wish to move forward on it feel free to close this one.

### CODE REVIEW

File: numpy/core/src/multiarray/iterators.c [hunk 1]
```diff
@@ -1242,232 +1242,183 @@ PyArray_Broadcast(PyArrayMultiIterObject *mit)
...
+            if (arr == NULL) {
+                goto fail;
             }
+            it = (PyArrayIterObject *)PyArray_IterNew(arr);
+            if (it == NULL) {
+                Py_DECREF(arr);
+                goto fail;
+            }
+            multi->iters[multi->numiter++] = it;
             Py_DECREF(arr);
```
Comment (seberg) [inline numpy/core/src/multiarray/iterators.c]:
```suggestion
```

File: numpy/core/src/multiarray/iterators.c [hunk 2]
```diff
@@ -1242,232 +1242,183 @@ PyArray_Broadcast(PyArrayMultiIterObject *mit)
...
+            }
+        }
+        else if (multi->numiter < NPY_MAXARGS) {
+            arr = PyArray_FromAny(obj, NULL, 0, 0, 0, NULL);
+            if (arr == NULL) {
+                goto fail;
             }
+            it = (PyArrayIterObject *)PyArray_IterNew(arr);
+            if (it == NULL) {
+                Py_DECREF(arr);
```
Comment (seberg) [inline numpy/core/src/multiarray/iterators.c]:
```suggestion
```

File: numpy/core/src/multiarray/iterators.c [hunk 3]
```diff
@@ -1242,232 +1242,183 @@ PyArray_Broadcast(PyArrayMultiIterObject *mit)
...
+                }
+                multi->iters[multi->numiter++] = it;
+            }
+        }
+        else if (multi->numiter < NPY_MAXARGS) {
+            arr = PyArray_FromAny(obj, NULL, 0, 0, 0, NULL);
+            if (arr == NULL) {
+                goto fail;
             }
+            it = (PyArrayIterObject *)PyArray_IterNew(arr);
```
Comment (seberg) [inline numpy/core/src/multiarray/iterators.c]:
```suggestion
            it = (PyArrayIterObject *)PyArray_IterNew(arr);
            Py_DECREF(arr);
```
I think you can move the DECREF up, sorry I guess those suggestions will mess things up again... (I hope they enable suggestion to batch when in review mode...)

**Review by seberg — APPROVED** [review]
tl;dr: Makes the C-Api a bit more general, but it is a super corner case and seems OK, and not branching is much nicer.

It seems to me that the C-API functions `PyArray_MultiIterNew` and `PyArray_MultiIterFromObjects` previously only accepted arrays, while the Py-API also accepted iterator objects. The whole thing seems a bit strange to me, but on the other hand, it seems also harmless to generalize (and we could deprecate it at some point if we want to). It seem impossible that someone relies on MultiIterObjects getting broadcast when using the C-API. Those are two very very unlikely things together.

File: numpy/core/src/multiarray/iterators.c [hunk 4]
```diff
@@ -1242,232 +1242,183 @@ PyArray_Broadcast(PyArrayMultiIterObject *mit)
...
+{
+    return PyErr_Format(PyExc_ValueError,
+                        "Need at least 1 and at most %d "
+                        "array objects.", NPY_MAXARGS);
+}
+
+/*
+ * Common implementation for all PyArrayMultiIterObject constructors.
+ * This function takes a pointer to an array of at most NPY_MAXARGS
+ * PyObject pointers, NULL terminated unless completely full, which
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/iterators.c]: Is there any point null terminating here? I'd be inclined to just pass the length separately, and change line 1272 to `for (i = 0; i < n_args; i++)`. Note this loop does not care about `MAX_ARGS`, that's already covered by the `multi->numiter < NPY_MAXARGS` within the loop.

File: numpy/core/src/multiarray/iterators.c [hunk 5]
```diff
@@ -1242,232 +1242,183 @@ PyArray_Broadcast(PyArrayMultiIterObject *mit)
...
-    if (PyArray_Broadcast(multi) < 0) {
-        goto fail;
-    }
-    PyArray_MultiIter_RESET(multi);
-    return (PyObject *)multi;
-
- fail:
-    Py_DECREF(multi);
-    return NULL;
+    return multiiter_new_impl(args_impl);
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/iterators.c]: If you had a length argument here, you avoid a copy with something like
```
PyObject *fast_seq = PySequence_Fast(args, "");  // needed for pypy
assert(fast_seq);
PyObject *ret = multiiter_new_impl(PySequenceFast_ITEMS(fast_seq), PySequenceFast_SIZE(fast_seq));
Py_DECREF(fast_seq);
return ret;
```

**Comment by mattip** [comment]
Updated with all requested changes, including using `PySequence_Fast`

**Review by seberg — APPROVED** [review]
OK, looks good to me. Could you not squash it right away next time maybe? Makes looking at followup changes a bit easier (not sure it would have helped here).

There is a merge conflict unfortunately.

**Comment by seberg** [comment]
Forced pushed a hopefully good rebase (it was unfortunate I/we merged the other one first), undos some of the error message changes, but I do not think that matters. But it may need at least one more glance through now unfortunately.

**Comment by mattip** [comment]
@seberg ping

**Review by seberg — APPROVED** [review]
Looks good to me, this undid the mnore detailed error messages now, but that higher detail did not give much anyway (negative number is implausible, so I think it is OK to not distinguish the two).

Cannot see anything wrong or any merge issue that arose. Matti, go ahead and merge just in case you wanted to have a glance because of the merge. Otherwise I will merge in a day or so.
